### PR TITLE
Pt 154245674 oracle api

### PR DESCRIPTION
**Pull request by hanssv**
WebSocket API and life cycle example using the API

### CODE REVIEW

**Comment by hanssv** [comment]
interaction_id and query_id are used in a confusing way - will try to make a patch for this...

File: oracles/oracle_ws_api.md [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+you need a node per account - obviously this is a limitation we should try to
+lift as soon as possible.
+
+## Detailed example - Oracle life cycle
+
+For simplicity, we only work with a single node (`localhost/127.0.0.1`) in this
+example, but it should be straightforward to split the example into one node
+running the oracle and other nodes querying that oracle. In the default node
+configuration the websocket interface is at *port 3014* - i.e. connecting to
+`ws://127.0.0.1:3114/websocket` open a websocket connection.
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/open/opens/

File: oracles/oracle_ws_api.md [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+In order to work through the example we also need the (Base58Check-encoded)
+public key of the node. This is easily retrieved from the running node:
+```
+~/epoch/node: curl http://127.0.0.1:3113/v1/account/pub-key
+{"pub_key":"ak$3jzZyCLFtHVD7yVdEhGJFM3LjeXrKqWxnHbCYzhnrrR4DkdFtaJuxQvrR8VbbXExDPkCHFAei5q969JA6EayQpb8z5C3Mf"}
+```
+
+On a freshly started node, connect a websocket. (Here we use
+[WSCat](https://github.com/websockets/wscat) for the websocket connection.)
+Upon connection the client automatically subscribe to the _mined_block_-event,
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/subscribe/subscribes/

File: oracles/oracle_ws_api.md [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+i.e. everytime the node mine a new block an event is pushed to the client. We
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/mine/mines/

File: oracles/oracle_ws_api.md [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+```
+
+Once the account has a positive balance we can post an _"Oracle register transaction"_:
+```
+...
+> {"target":"oracle", "action":"register", "payload":{"type":"OracleRegisterTxObject", "vsn":1, "account":"ak$3jzZyCLFtHVD7yVdEhGJFM3LjeXrKqWxnHbCYzhnrrR4DkdFtaJuxQvrR8VbbXExDPkCHFAei5q969JA6EayQpb8z5C3Mf", "query_format":"the query spec", "response_format":"the response spec", "query_fee":4, "ttl":{"type":"delta", "value":50}, "fee":5}}
+< {"action":"register","origin":"oracle","payload":{"result":"ok","oracle_id":"ok$3jzZyCLFtHVD7yVdEhGJFM3LjeXrKqWxnHbCYzhnrrR4DkdFtaJuxQvrR8VbbXExDPkCHFAei5q969JA6EayQpb8z5C3Mf"}}
+```
+
+The register transaction use the same format for the payload as the HTTP API
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/use/uses/

File: oracles/oracle_ws_api.md [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+Oracle:
+```
+...
+< {"action":"mined_block","origin":"miner","payload":{"height":2,"hash":"bh$2UuiuAHW8bmRkcoMqP2Mcj5ZrxvGzjhQRhictByxf4AhcyF4q"}}
+> {"target":"oracle","action":"subscribe", "payload":{"type":"query","oracle_id":"ok$3jzZyCLFtHVD7yVdEhGJFM3LjeXrKqWxnHbCYzhnrrR4DkdFtaJuxQvrR8VbbXExDPkCHFAei5q969JA6EayQpb8z5C3Mf"}}
+< {"action":"subscribe","origin":"oracle","payload":{"result":"ok","subscribed_to":{"oracle_id":"ok$3jzZyCLFtHVD7yVdEhGJFM3LjeXrKqWxnHbCYzhnrrR4DkdFtaJuxQvrR8VbbXExDPkCHFAei5q969JA6EayQpb8z5C3Mf","type":"query"}}}
+```
+
+We subscribe for `"type":"query"`, i.e. we want to get notified when the oracle
+with the given id receive a query. Of course we want to test this, so let us
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/receive/receives/

File: oracles/oracle_ws_api.md [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+```
+
+Again we use exactly the same format for the payload as the HTTP API. And also,
+similar to the oracle registration transaction, the _transaction fee_ depends
+on the _TTL_ **and** the oracle query fee. The base query transaction fee is 2,
+we set the oracle query fee to 4 when registering, and the TTL accounts for 1
+per 1000 blocks (just a preliminary value to test the concept). I.e. the cost
+for our test transaction is 7.
+
+Not much will happen until the node mine another block (and the query
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/mine/mines/

File: oracles/oracle_ws_api.md [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+rather than Bob...)
+
+```
+...
+> {"target":"oracle","action":"subscribe", "payload":{"type":"query","oracle_id":"ok$3jzZyCLFtHVD7yVdEhGJFM3LjeXrKqWxnHbCYzhnrrR4DkdFtaJuxQvrR8VbbXExDPkCHFAei5q969JA6EayQpb8z5C3Mf"}}
+< {"action":"subscribe","origin":"oracle","payload":{"result":"ok","subscribed_to":{"oracle_id":"ok$3jzZyCLFtHVD7yVdEhGJFM3LjeXrKqWxnHbCYzhnrrR4DkdFtaJuxQvrR8VbbXExDPkCHFAei5q969JA6EayQpb8z5C3Mf","type":"query"}}}
+```
+
+We subscribe for `"type":"response"`, i.e. we want to get notified when the
+oracle answer the query with the given id. Of course we want to test this as
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/answer/answers/

File: oracles/oracle_ws_api.md [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+If we then wait for another block to be mined (and the response transaction to
+be added to the chain), we see that we receive an event reflecting the query
+response:
+```
+...
+< {"action":"mined_block","origin":"miner","payload":{"height":6,"hash":"bh$i9P6dgcihe47Kwm6ZvTDb84HPGWr147KPviatJ5QXntHrBY1m"}}
+< {"action":"new_oracle_response","origin":"node","payload":{"interaction_id":"oi$4RZoMEkm8QuuhJiiq53dd5pE4VstCthYRjBHgUKdhAhe7rLEr","response":"I am fine, thanks!"}}
+```
+
+This conclude the walk through of the oracle life cycle example, which also
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/conclude the walk through/concludes the walkthrough/

File: oracles/oracle_ws_api.md [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+exercise the websocket API for Oracles fully.
```
Comment (knarz) [inline oracles/oracle_ws_api.md]: s/exercise/exercises/

File: oracles/oracles.md [hunk 10]
```diff
@@ -16,6 +16,8 @@
 
 ## [Oracle life cycle examples](./oracle_life_cycle.md)
 
+## [Oracle WebSocket API](./oracle_ws_api.md)
```
Comment (lucafavatella) [inline oracles/oracles.md]: When we spoke today I understood we wanted to keep epoch API in this repo though segregated somehow. I would propose last section in this document `## epoch node API` and separate folder `./epoch_api` containing API docs.

File: oracles/oracle_ws_api.md [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+| ---- | ---- | ----------- | -------- |
+| type | `delta` |  | Yes |
+| value | number |  | Yes |
+
+
+**Current limitation:** Since there is currently no interface for offline signing
+(all signing is done by a node - and that node is associated with exactly one
+account) it is somewhat limited what you can do with oracles. I.e. in practice
+you need a node per account - obviously this is a limitation we should try to
+lift as soon as possible.
```
Comment (lucafavatella) [inline oracles/oracle_ws_api.md]: This looks an action. Is this tracked somewhere and discussed?

File: oracles/oracle_ws_api.md [hunk 12]
```diff
@@ -0,0 +1,427 @@
+[back](./oracles.md)
```
Comment (lucafavatella) [inline oracles/oracle_ws_api.md]: May I suggest splitting this document in two parts:
* A document that acts as a surrogate for the Swagger schema for websocket
* A document that documents to intended usage of API (the portion "For simplicity, we only work with a single node...")

File: oracles/oracle_ws_api.md [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+
+### Response
+ * **origin:** `oracle`
+ * **action:** `query`
+ * **payload:**
+
+  | Name | Type | Description | Required |
+  | ---- | ---- | ----------- | -------- |
+  | result | string | `ok` or error reason | Yes |
+  | query_id | string | Hash for created query | Yes on success |
```
Comment (lucafavatella) [inline oracles/oracle_ws_api.md]: Is the query_id an id per-oracle or across all oracles?

Comment (lucafavatella) [reply oracles/oracle_ws_api.md]: and is user meant to be able to extract oracle from query (I think not needed)?

File: oracles/oracle_ws_api.md (hunk 11 above)
Comment (knarz) [reply oracles/oracle_ws_api.md]: I created this task a while ago, which is related: https://www.pivotaltracker.com/story/show/153451598

File: oracles/oracle_ws_api.md (hunk 13 above)
Comment (hanssv) [reply oracles/oracle_ws_api.md]: query_id is unique (across all oracles) 

no, you can't extract the oracle_id from the query_id

**Comment by hanssv** [comment]
Introduced epoch_api and put the more formal description of the Oracle WS API there. Kept the _intended usage_ part under Oracles...

Also updated the State tree description for Oracles.

File: oracles/oracle_state_tree.md [hunk 14]
```diff
@@ -58,16 +60,13 @@ has the benefits:
...
-If the oracle interaction has not been given a response, the poster of
+If the oracle query has not been given a response, the poster of
 the query should be refunded the oracle query fee. If the oracle has
 responded, the oracle was already given the funds at response time.
 
-### TODO:
-- Should the oracle state tree be a separate tree or
-  be joined with the accounts state tree?
-- Define the contents of the oracle interaction object.
+**NOTE:** The refunding has not yet been implemented
```
Comment (lucafavatella) [inline oracles/oracle_state_tree.md]: Thanks for the note.

Comment (hanssv) [reply oracles/oracle_state_tree.md]: It is also tracked by PT-#154292757

**Pull request merged**